File: Topological Sorting/9470번 - Strahler 순서/9470.py
import sys
from collections import deque

input = sys.stdin.readline

def solve():
    ans = 0
    Q = deque()
    for i in range(1, m + 1):
        if not degree[i]:
            strahler[i][0] = 1
            Q.append(i)
            
    while Q:
        cur = Q.popleft()
        ans = max(ans, strahler[cur][0])
        
        for ncur in river[cur]:
            if strahler[ncur][0] < strahler[cur][0]:
                strahler[ncur][0] = strahler[cur][0]
                strahler[ncur][1] = False
            elif strahler[ncur][0] == strahler[cur][0] and not strahler[ncur][1]:
                strahler[ncur][0] += 1
                strahler[ncur][1] = True
                
            degree[ncur] -= 1
            if degree[ncur] == 0:
                Q.append(ncur)
                
    print(k, ans)

if __name__ == '__main__':
    tc = int(input())
    while tc:
        k, m, p = map(int, input().split())
        river = list([] for _ in range(m + 1))
        degree = list([0] * (m + 1))
        strahler = list([0, False] for _ in range(m + 1))
        for _ in range(p):
            a, b = map(int, input().split())
            river[a].append(b)
            degree[b] += 1
        solve()
        
        tc -= 1

File: Topological Sorting/9470번 - Strahler 순서/9470.cpp
// 9470번: Strahler 순서
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define MAX 1001

int k, m, p;
vector<int> river[MAX];
int degree[MAX] = { 0, };
int strahler[2][MAX] = { 0, };

void input() {
    for(int i = 0; i < MAX; i++) river[i].clear();
    fill_n(&degree[0], MAX, 0);
    fill_n(&strahler[0][0], 2* MAX, 0);
    
    cin >> k >> m >> p;
    for(int i = 0, a, b; i < p; i++) {
        cin >> a >> b;
        river[a].push_back(b);
        degree[b]++;
    }
}

void solve() {
    queue<int> Q;
    int ans = 0;
    
    for(int i = 1; i <= m; i++) {
        if(!degree[i]) {
            Q.push(i);
            strahler[0][i] = 1;
        }
    }
    
    while(!Q.empty()) {
        int cur = Q.front(); Q.pop();
        ans = max(ans, strahler[0][cur]);
        
        for(int i = 0; i < river[cur].size(); i++) {
            int ncur = river[cur][i];
            if(strahler[0][ncur] < strahler[0][cur]) {
                strahler[0][ncur] = strahler[0][cur];
                strahler[1][ncur] = 0;
            }
            else if(strahler[0][ncur] == strahler[0][cur] && !strahler[1][ncur]) {
                strahler[0][ncur]++;
                strahler[1][ncur] = 1;    
            }
            
            if(--degree[ncur] == 0) Q.push(ncur);
        }
    }
    
    cout << k << ' ' << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int tc;
    cin >> tc;
    while(tc--) {
        input();
        solve();
    }
    
    return 0;
}